### PR TITLE
[OTLP] Export LogRecord.CategoryName as InstrumentationScope name

### PR DESCRIPTION
**Pull request by vishweshbankwar**
May fix https://github.com/open-telemetry/opentelemetry-dotnet/issues/4867

## Changes

This PR sets the [InstrumentationScope](https://github.com/open-telemetry/opentelemetry-dotnet/blob/3c2bb7c93dd2e697636479a1882f49bb0c4a362e/src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/opentelemetry/proto/logs/v1/logs.proto#L64-L75) `name` field under ScopeLogs to `LogRecord.CategoryName`. With this change exported logs will be grouped by the `LogRecord.CategoryName`.

As an example consider following scenario

```
var logger1 = loggerFactory.CreateLogger("ProgramA");

var logger2 = loggerFactory.CreateLogger("ProgramB");

 

logger1.FoodPriceChanged("artichoke", 9.99);

logger2.FoodPriceChanged("grapes", 9.99);
```

OTLP payload will look like below: The logs are grouped by CategoryName

```json
{
  "resourceLogs": [
    {
      "resource": {
        "attributes": [
          {
            "key": "telemetry.sdk.name",
            "value": {
              "stringValue": "opentelemetry"
            }
          },
          {
            "key": "telemetry.sdk.language",
            "value": {
              "stringValue": "dotnet"
            }
          },
          {
            "key": "telemetry.sdk.version",
            "value": {
              "stringValue": "1.6.1-alpha.0.53"
            }
          },
          {
            "key": "service.name",
            "value": {
              "stringValue": "unknown_service:getting-started-console"
            }
          }
        ]
      },
      "scopeLogs": [
        {
          "scope": {
            "name": `ProgramA`
          },
          "logRecords": [
            {
              "timeUnixNano": "1697049028570433100",
              "severityNumber": "SEVERITY_NUMBER_INFO",
              "severityText": "Information",
              "body": {
                "stringValue": "Food `{name}` price changed to `{price}`."
              },
              "attributes": [
                {
                  "key": "name",
                  "value": {
                    "stringValue": "artichoke"
                  }
                },
                {
                  "key": "price",
                  "value": {
                    "doubleValue": 9.99
                  }
                }
              ],
              "observedTimeUnixNano": "1697049028570433100"
            }
          ]
        },
        {
          "scope": {
            "name": `ProgramB`
          },
          "logRecords": [
            {
              "timeUnixNano": "1697049028743206200",
              "severityNumber": "SEVERITY_NUMBER_INFO",
              "severityText": "Information",
              "body": {
                "stringValue": "Food `{name}` price changed to `{price}`."
              },
              "attributes": [
                {
                  "key": "name",
                  "value": {
                    "stringValue": "grapes"
                  }
                },
                {
                  "key": "price",
                  "value": {
                    "doubleValue": 9.99
                  }
                }
              ],
              "observedTimeUnixNano": "1697049028743206200"
            }
          ]
        }
      ]
    }
  ]
}
```

## Merge requirement checklist

* [ ] [CONTRIBUTING](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/CONTRIBUTING.md) guidelines followed (nullable enabled, static analysis, etc.)
* [x] Unit tests added/updated
* [x] Appropriate `CHANGELOG.md` files updated for non-trivial changes
* [ ] Changes in public API reviewed (if applicable)

### CODE REVIEW

**Comment by CodeBlanch** [comment]
I have one bit of feedback. There is actually support for instrumentation scope on `LogRecord` today! It is `LogRecord.Logger.Name` (also version there). I would vote for the SDK to set `LogRecord.Logger` based on the `categoryName` for the `ILogger` path. Then we can update OTLP to use that. What that would accomplish is if people use the experimental log bridge stuff it will also work correctly. But if you want to tackle that later, fine by me.

(I think this would make @alanwest happy because it would remove any knowledge of "category name" from OTLP 😄)

**Comment by vishweshbankwar** [comment]
> I have one bit of feedback. There is actually support for instrumentation scope on `LogRecord` today! It is `LogRecord.Logger.Name` (also version there). I would vote for the SDK to set `LogRecord.Logger` based on the `categoryName` for the `ILogger` path. Then we can update OTLP to use that. What that would accomplish is if people use the experimental log bridge stuff it will also work correctly. But if you want to tackle that later, fine by me.
> 
> (I think this would make @alanwest happy because it would remove any knowledge of "category name" from OTLP 😄)

That part requires experimental stuff. I would like to keep that one separate from this.

**Comment by CodeBlanch** [comment]
> That part requires experimental stuff.

Not true! Otlp will see it either public or internal but everything is there and the logic is the same either way. Whether or not the bridge api is available for other things to set it depends on experimental stuff but the SDK + LogRecord portion is always there.

**Comment by vishweshbankwar** [comment]
> > That part requires experimental stuff.
> 
> Not true! Otlp will see it either public or internal but everything is there and the logic is the same either way. Whether or not the bridge api is available for other things to set it depends on experimental stuff but the SDK + LogRecord portion is always there.

I was not aware of this: https://github.com/open-telemetry/opentelemetry-dotnet/blob/59f1d2ce8dbea9ead590a8e184780792c315d4d8/src/OpenTelemetry/AssemblyInfo.cs#L30-L32

I can do that as a separate change. If that is ok?

File: src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/OtlpLogRecordTransformer.cs [hunk 1]
```diff
@@ -49,21 +54,69 @@ public OtlpLogRecordTransformer(SdkLimitOptions sdkLimitOptions, ExperimentalOpt
...
+                if (!this.logsByCategory.TryGetValue(logRecord.CategoryName, out var scopeLogs))
+                {
+                    scopeLogs = this.GetLogListFromPool(logRecord.CategoryName);
+                    scopeLogs.LogRecords.Add(otlpLogRecord);
+                    this.logsByCategory.Add(logRecord.CategoryName, scopeLogs);
+                    resourceLogs.ScopeLogs.Add(scopeLogs);
+                }
+                else
+                {
+                    scopeLogs.LogRecords.Add(otlpLogRecord);
```
Comment (utpilla) [inline src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/OtlpLogRecordTransformer.cs:73]: nit: We can simplify this to remove the `else` block. This line is present in both `if` and `else` blocks.

File: src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/OtlpLogRecordTransformer.cs [hunk 2]
```diff
@@ -41,6 +44,9 @@ public OtlpLogRecordTransformer(SdkLimitOptions sdkLimitOptions, ExperimentalOpt
         OtlpResource.Resource processResource,
         in Batch<LogRecord> logRecordBatch)
     {
+        // https://github.com/open-telemetry/opentelemetry-dotnet/issues/4943
```
Comment (utpilla) [inline src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/OtlpLogRecordTransformer.cs:47]: Add the word "TODO" to make it show up on search results for it.

File: src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/OtlpLogRecordTransformer.cs [hunk 3]
```diff
@@ -49,21 +55,66 @@ public OtlpLogRecordTransformer(SdkLimitOptions sdkLimitOptions, ExperimentalOpt
         };
         request.ResourceLogs.Add(resourceLogs);
 
-        var scopeLogs = new OtlpLogs.ScopeLogs();
-        resourceLogs.ScopeLogs.Add(scopeLogs);
+        logsByCategory.Clear();
```
Comment (utpilla) [inline src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/OtlpLogRecordTransformer.cs:58]: This can be removed.

**Comment by CodeBlanch** [comment]
@vishweshbankwar 

> I can do that as a separate change. If that is ok?

Works for me!

File: src/OpenTelemetry.Exporter.OpenTelemetryProtocol/OtlpLogExporter.cs [hunk 4]
```diff
@@ -90,10 +90,10 @@ public override ExportResult Export(in Batch<LogRecord> logRecordBatch)
         // Prevents the exporter's gRPC and HTTP operations from being instrumented.
         using var scope = SuppressInstrumentationScope.Begin();
 
+        var request = this.otlpLogRecordTransformer.BuildExportRequest(this.ProcessResource, logRecordBatch);
```
Comment (utpilla) [inline src/OpenTelemetry.Exporter.OpenTelemetryProtocol/OtlpLogExporter.cs:93]: Why move this outside of the `try` block?

File: src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/OtlpLogRecordTransformer.cs [hunk 5]
```diff
@@ -49,21 +55,64 @@ public OtlpLogRecordTransformer(SdkLimitOptions sdkLimitOptions, ExperimentalOpt
...
+
                 scopeLogs.LogRecords.Add(otlpLogRecord);
             }
         }
 
         return request;
     }
 
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    internal void Return(OtlpCollector.ExportLogsServiceRequest request)
```
Comment (utpilla) [inline src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/OtlpLogRecordTransformer.cs:78]: Stay consistent with Traces and Metrics and change this to an extension method?

Comment (vishweshbankwar) [reply src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/OtlpLogRecordTransformer.cs:78]: Planning on doing same refactor for traces and metrics. will remove the static classes.

File: src/OpenTelemetry.Exporter.OpenTelemetryProtocol/OtlpLogExporter.cs (hunk 4 above)
Comment (vishweshbankwar) [reply src/OpenTelemetry.Exporter.OpenTelemetryProtocol/OtlpLogExporter.cs]: Not intentional. Fixed it

File: test/OpenTelemetry.Exporter.OpenTelemetryProtocol.Tests/OtlpLogExporterTests.cs [hunk 6]
```diff
@@ -1242,6 +1243,51 @@ public void AddOtlpLogExporterLogRecordProcessorOptionsTest(ExportProcessorType
...
+
+        var resourceBuilder = ResourceBuilder.CreateEmpty();
+        var processResource = resourceBuilder.Build().ToOtlpResource();
+
+        var request = logRecordTransformer.BuildExportRequest(processResource, batch);
+
+        Assert.Single(request.ResourceLogs);
+
+        Assert.Equal("OtlpLogExporterTests-A", request.ResourceLogs[0].ScopeLogs.First().Scope.Name);
+        Assert.Equal("OtlpLogExporterTests-B", request.ResourceLogs[0].ScopeLogs.Last().Scope.Name);
```
Comment (utpilla) [inline test/OpenTelemetry.Exporter.OpenTelemetryProtocol.Tests/OtlpLogExporterTests.cs:1276]: Add asserts to check that the exporter has indeed put the logs with the expected category under the scopes.

Comment (vishweshbankwar) [reply test/OpenTelemetry.Exporter.OpenTelemetryProtocol.Tests/OtlpLogExporterTests.cs]: done

**Review by utpilla — APPROVED** [review]

**Review by alanwest — APPROVED** [review]

**Pull request merged**